src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -331,16 +341,33 @@ static void verifyColumnMasterKeyMetadata(SQLServerConnection connection, SQLSer
      */
     static SqlAuthenticationToken getManagedIdentityCredAuthToken(String resource,
             String managedIdentityClientId) throws SQLServerException {
-        ManagedIdentityCredential mic = null;
 
         if (logger.isLoggable(java.util.logging.Level.FINEST)) {
             logger.finest("Getting Managed Identity authentication token for: " + managedIdentityClientId);
         }
 
-        if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
-            mic = new ManagedIdentityCredentialBuilder().clientId(managedIdentityClientId).build();
-        } else {
-            mic = new ManagedIdentityCredentialBuilder().build();
+        String key = getHashedSecret(new String[]{managedIdentityClientId, ManagedIdentityCredential.class.getSimpleName()});
+        ManagedIdentityCredential mic = (ManagedIdentityCredential) getCredentialFromCache(key);
+
+        if (null == mic) {
+            LOCK.lock();
+
+            try {
+                mic = (ManagedIdentityCredential) getCredentialFromCache(key);
+                if (null == mic) {
+                    ManagedIdentityCredentialBuilder micBuilder = new ManagedIdentityCredentialBuilder();
+
+                    if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
+                        mic = micBuilder.clientId(managedIdentityClientId).build();
+                        addCredentialToCache(key, mic);
+                    } else {
+                        mic = micBuilder.build();
+                        addCredentialToCache(key, mic);
+                    }
+                }

```

Comment:
```suggestion
                    if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
                        mic = micBuilder.clientId(managedIdentityClientId).build();
                    } else {
                        mic = micBuilder.build();
                    }
                    Credential credential = new Credential(mic);
                    CREDENTIAL_CACHE.put(key, credential);
                }

```
`addCredentialToCache` can be put outside the if/else. Also is the `addCredentialToCache` method needed? It's only two lines. I think its fine replacing this instance and the instance on 448 with the method contents and removing the method.